src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1484,6 +1484,17 @@ private BoundAssignmentOperator BindAssignment(
 
             if (op1.Type is { } lhsType && !lhsType.IsErrorType())
             {
+                Debug.Assert(!op1.NeedsToBeConverted()
+                    // These are only required to be converted in Debug to increase confidence that we've called BindToTypeForErrorRecovery or BindToNaturalType
+                    // on the operand. We're calling BindToTypeForErrorRecovery below, so we're good.
+                    || op1 is BoundParameter or BoundLocal
+                    || op1.HasAnyErrors);
+
+                if (op1.HasAnyErrors)
```

Comment:
>HasAnyErrors [](http://example.com/codeflow?start=24&length=12)

This check seems too broad since the error could in a nested expression with no error on `op1` directly. In short, it seems surprising we're binding for error recovery in that situation.

If we're just trying to ensure `op1` is converted, would it make sense to do that explicitly?
```csharp
if (op1.NeedsToBeConverted())
{
    op1 = BindToNaturalType(op1, BindingDiagnosticBag.Discard, reportNoTargetType: false);
}
```

Reply:
My original approach did effectively this. Aleksey was concerned, rightly so, that discarding diagnostics here is probably not safe, as there is no guarantee that errors have been reported for `op1`. I do not think that it is surprising that we bind for error recovery in this scenario when we know that there are errors; in such a scenario, we simply want to be able to give the most precise info that we can.

Reply:
> discarding diagnostics here is probably not safe, as there is no guarantee that errors have been reported for `op1`.

In that case, we could pass in the `diagnostics` instance.

> I do not think that it is surprising that we bind for error recovery in this scenario when we know that there are errors

`HasAnyErrors` indicates there is an error somewhere beneath `op1`, but the outermost node `op1` might be otherwise valid.

Reply:
If you would prefer, I can switch to just checking `HasErrors`, rather than `HasAnyErrors`, so it's only on the node itself.